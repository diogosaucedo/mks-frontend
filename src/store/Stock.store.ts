import { PayloadAction, createSlice } from '@reduxjs/toolkit';

import { ICardProduct } from '@/interfaces';

const stock = createSlice({
  name: 'stock',
  initialState: {
    counter: 0,
    total: 0,
    visible: false,
    products: [] as never | ICardProduct[],
  },
  reducers: {
    toggleVisible(state) {
      state.visible = !state.visible;
    },
    addProduct(state, action: PayloadAction<ICardProduct>) {
      if (
        state.products.find(
          (item: ICardProduct) => item.id == action.payload.id,
        )
      ) {
      } else {
        state.products.push(action.payload);
      }
    },
    removeProduct(state, action: PayloadAction<number>) {
      state.products = state.products.filter(
        (product) => product.id !== action.payload,
      );
    },
    counter(state) {
      state.counter = state.products.length;
    },
    getTotal(state) {
      state.total = state.products.reduce(
        (acc, item) => acc + item.amount * item.price,
        0,
      );
    },
    incrementAmountProduct(state, action: PayloadAction<number>) {
      state.products = state.products.map((product) => {
        if (product.id === action.payload) {
          return { ...product, amount: product.amount + 1 };
        } else {
          return product;
        }
      });
    },
    decrementAmountProduct(state, action: PayloadAction<number>) {
      state.products = state.products.map((product) => {
        if (product.id === action.payload && product.amount > 1) {
          return { ...product, amount: product.amount - 1 };
        } else {
          return product;
        }
      });
    },
  },
});

export const {
  toggleVisible,
  addProduct,
  counter,
  removeProduct,
  getTotal,
  incrementAmountProduct,
  decrementAmountProduct,
} = stock.actions;
export default stock.reducer;
