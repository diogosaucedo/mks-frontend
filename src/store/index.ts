import { configureStore } from '@reduxjs/toolkit'
import stockStore from './Stock.store'

export const store = configureStore({
  reducer: {
    stock: stockStore
  },
})

// Infer the `RootState` and `AppDispatch` types from the store itself
export type RootState = ReturnType<typeof store.getState>
// Inferred type: {posts: PostsState, comments: CommentsState, users: UsersState}
export type AppDispatch = typeof store.dispatch
