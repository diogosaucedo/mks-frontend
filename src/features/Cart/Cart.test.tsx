import React from 'react';
import { render, screen } from '@testing-library/react';
import { useDispatch, useSelector } from 'react-redux';
import { toggleVisible } from '@/store/Stock.store';
import { Cart } from '.';

jest.mock('react-redux', () => ({
  useDispatch: jest.fn(),
  useSelector: jest.fn(),
}));

describe('Cart feature', () => {
  const dispatch = jest.fn();
  const mockSelector = jest.fn().mockReturnValue({
    visible: true,
    products: [
      { id: 1, photo: 'photo1.jpg', name: 'Product 1', price: 9.99 },
      { id: 2, photo: 'photo2.jpg', name: 'Product 2', price: 19.99 },
    ],
    total: 29.98,
  });

  beforeEach(() => {
    (useDispatch as jest.Mock).mockReturnValue(dispatch);
    (useSelector as jest.Mock).mockImplementation(mockSelector);
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should render Cart component', () => {
    render(<Cart />);

    expect(screen.getByText('Carrinho de Compras')).toBeInTheDocument();
    expect(screen.getByText('Finalizar Compra')).toBeInTheDocument();
    expect(screen.getByText('R$29.98')).toBeInTheDocument();
  });

  it('should match snapshot', () => {
    const { asFragment } = render(<Cart />);

    expect(asFragment()).toMatchSnapshot();
  });

  it('should call toggleVisible when clicking on close button', () => {
    render(<Cart />);

    screen.getByRole('button', { name: 'icon close' }).click();
    expect(dispatch).toHaveBeenCalledWith(toggleVisible());
  });
});
