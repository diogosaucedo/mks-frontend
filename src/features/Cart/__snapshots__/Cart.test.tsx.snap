// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Cart /> should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="sc-hLBbgP bksghu"
  >
    <div
      class="sc-iBYQkv ftBjEV"
      width="100%"
    >
      <div
        class="sc-iBYQkv JgEpM"
        width="90%"
      >
        <div
          class="sc-bcXHqe fiMeFo"
        >
          <h2
            class="sc-eDvSVe hOkahJ"
            color="#FFFFFF"
            font-size="27px"
            type="h2"
            width="180px"
          >
            Carrinho de Compras
          </h2>
          <button
            class="sc-gswNZR ccNznE"
            height="38px"
            width="38px"
          >
            <img
              alt="icon close"
              class="sc-pyfCe kPsozc"
              src="x.svg"
              width="15px"
            />
          </button>
        </div>
        <div
          class="sc-iBYQkv kFWzbU"
        >
          <div
            class="sc-ipEyDJ hYSMdb sc-csuSiG kHiHdU"
          >
            <div
              class="sc-bcXHqe cojVzm"
            >
              <div
                class="sc-iBYQkv eholvE"
              >
                <img
                  alt="Product 1"
                  class="sc-pyfCe ljnZwJ"
                  height="60px"
                  src="photo1.jpg"
                  width="fit-content"
                />
                <p
                  class="sc-dkrFOg fDejt"
                  font-size="13px"
                  width="100px"
                >
                  Product 1
                </p>
              </div>
              <div
                class="sc-bcXHqe hagkwD"
              >
                <div
                  class="sc-bcXHqe koBore"
                >
                  <p
                    class="sc-dkrFOg fvHszv"
                    font-size="5px"
                  >
                    Qtd:
                  </p>
                  <div
                    class="sc-bcXHqe jHqwTT"
                  >
                    <button
                      class="sc-gswNZR hFwvBQ"
                      font-size="12px"
                      height="19px"
                      width="15px"
                    >
                      -
                    </button>
                    <input
                      class="sc-gKPRtg bKYIac"
                      disabled=""
                      font-size="12px"
                      height="19px"
                      type="text"
                      value=""
                      width="20px"
                    />
                    <button
                      class="sc-gswNZR hFwvBQ"
                      font-size="12px"
                      height="19px"
                      width="15px"
                    >
                      +
                    </button>
                  </div>
                </div>
                <p
                  class="sc-dkrFOg dNFUVo sc-jSUZER QifgC"
                  font-size="14px"
                  font-weight="700"
                >
                  R$9
                </p>
              </div>
            </div>
            <button
              class="sc-gswNZR cnmGuK"
              height="18px"
              transform="translate(10px, -45px)"
              width="18px"
            >
              <p
                class="sc-dkrFOg harcBR"
                color="#FFFFFF"
                font-size="8px"
              >
                X
              </p>
            </button>
          </div>
          <div
            class="sc-ipEyDJ hYSMdb sc-csuSiG kHiHdU"
          >
            <div
              class="sc-bcXHqe cojVzm"
            >
              <div
                class="sc-iBYQkv eholvE"
              >
                <img
                  alt="Product 2"
                  class="sc-pyfCe ljnZwJ"
                  height="60px"
                  src="photo2.jpg"
                  width="fit-content"
                />
                <p
                  class="sc-dkrFOg fDejt"
                  font-size="13px"
                  width="100px"
                >
                  Product 2
                </p>
              </div>
              <div
                class="sc-bcXHqe hagkwD"
              >
                <div
                  class="sc-bcXHqe koBore"
                >
                  <p
                    class="sc-dkrFOg fvHszv"
                    font-size="5px"
                  >
                    Qtd:
                  </p>
                  <div
                    class="sc-bcXHqe jHqwTT"
                  >
                    <button
                      class="sc-gswNZR hFwvBQ"
                      font-size="12px"
                      height="19px"
                      width="15px"
                    >
                      -
                    </button>
                    <input
                      class="sc-gKPRtg bKYIac"
                      disabled=""
                      font-size="12px"
                      height="19px"
                      type="text"
                      value=""
                      width="20px"
                    />
                    <button
                      class="sc-gswNZR hFwvBQ"
                      font-size="12px"
                      height="19px"
                      width="15px"
                    >
                      +
                    </button>
                  </div>
                </div>
                <p
                  class="sc-dkrFOg dNFUVo sc-jSUZER QifgC"
                  font-size="14px"
                  font-weight="700"
                >
                  R$19
                </p>
              </div>
            </div>
            <button
              class="sc-gswNZR cnmGuK"
              height="18px"
              transform="translate(10px, -45px)"
              width="18px"
            >
              <p
                class="sc-dkrFOg harcBR"
                color="#FFFFFF"
                font-size="8px"
              >
                X
              </p>
            </button>
          </div>
        </div>
      </div>
      <div
        class="sc-iBYQkv tvFcD"
        width="100%"
      >
        <div
          class="sc-iBYQkv biqZAK"
          width="80%"
        >
          <p
            class="sc-dkrFOg kYzYoE"
            color="#FFFFFF"
            font-size="28px"
            font-weight="700"
          >
            Total:
          </p>
          <p
            class="sc-dkrFOg kYzYoE"
            color="#FFFFFF"
            font-size="28px"
            font-weight="700"
          >
            R$29.98
          </p>
        </div>
        <button
          class="sc-gswNZR cUQGLN"
          color="#FFFFFF"
          font-size="28px"
          height="97px"
          width="100%"
        >
          <p
            class="sc-dkrFOg cxIDJR"
            font-weight="700"
          >
            Finalizar Compra
          </p>
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Cart feature should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="sc-hLBbgP bksghu"
  >
    <div
      class="sc-iBYQkv ftBjEV"
      width="100%"
    >
      <div
        class="sc-iBYQkv JgEpM"
        width="90%"
      >
        <div
          class="sc-bcXHqe fiMeFo"
        >
          <h2
            class="sc-eDvSVe hOkahJ"
            color="#FFFFFF"
            font-size="27px"
            type="h2"
            width="180px"
          >
            Carrinho de Compras
          </h2>
          <button
            class="sc-gswNZR ccNznE"
            height="38px"
            width="38px"
          >
            <img
              alt="icon close"
              class="sc-pyfCe kPsozc"
              src="x.svg"
              width="15px"
            />
          </button>
        </div>
        <div
          class="sc-iBYQkv kFWzbU"
        >
          <div
            class="sc-ipEyDJ hYSMdb sc-csuSiG kHiHdU"
          >
            <div
              class="sc-bcXHqe cojVzm"
            >
              <div
                class="sc-iBYQkv eholvE"
              >
                <img
                  alt="Product 1"
                  class="sc-pyfCe ljnZwJ"
                  height="60px"
                  src="photo1.jpg"
                  width="fit-content"
                />
                <p
                  class="sc-dkrFOg fDejt"
                  font-size="13px"
                  width="100px"
                >
                  Product 1
                </p>
              </div>
              <div
                class="sc-bcXHqe hagkwD"
              >
                <div
                  class="sc-bcXHqe koBore"
                >
                  <p
                    class="sc-dkrFOg fvHszv"
                    font-size="5px"
                  >
                    Qtd:
                  </p>
                  <div
                    class="sc-bcXHqe jHqwTT"
                  >
                    <button
                      class="sc-gswNZR hFwvBQ"
                      font-size="12px"
                      height="19px"
                      width="15px"
                    >
                      -
                    </button>
                    <input
                      class="sc-gKPRtg bKYIac"
                      disabled=""
                      font-size="12px"
                      height="19px"
                      type="text"
                      value=""
                      width="20px"
                    />
                    <button
                      class="sc-gswNZR hFwvBQ"
                      font-size="12px"
                      height="19px"
                      width="15px"
                    >
                      +
                    </button>
                  </div>
                </div>
                <p
                  class="sc-dkrFOg dNFUVo sc-jSUZER QifgC"
                  font-size="14px"
                  font-weight="700"
                >
                  R$9
                </p>
              </div>
            </div>
            <button
              class="sc-gswNZR cnmGuK"
              height="18px"
              transform="translate(10px, -45px)"
              width="18px"
            >
              <p
                class="sc-dkrFOg harcBR"
                color="#FFFFFF"
                font-size="8px"
              >
                X
              </p>
            </button>
          </div>
          <div
            class="sc-ipEyDJ hYSMdb sc-csuSiG kHiHdU"
          >
            <div
              class="sc-bcXHqe cojVzm"
            >
              <div
                class="sc-iBYQkv eholvE"
              >
                <img
                  alt="Product 2"
                  class="sc-pyfCe ljnZwJ"
                  height="60px"
                  src="photo2.jpg"
                  width="fit-content"
                />
                <p
                  class="sc-dkrFOg fDejt"
                  font-size="13px"
                  width="100px"
                >
                  Product 2
                </p>
              </div>
              <div
                class="sc-bcXHqe hagkwD"
              >
                <div
                  class="sc-bcXHqe koBore"
                >
                  <p
                    class="sc-dkrFOg fvHszv"
                    font-size="5px"
                  >
                    Qtd:
                  </p>
                  <div
                    class="sc-bcXHqe jHqwTT"
                  >
                    <button
                      class="sc-gswNZR hFwvBQ"
                      font-size="12px"
                      height="19px"
                      width="15px"
                    >
                      -
                    </button>
                    <input
                      class="sc-gKPRtg bKYIac"
                      disabled=""
                      font-size="12px"
                      height="19px"
                      type="text"
                      value=""
                      width="20px"
                    />
                    <button
                      class="sc-gswNZR hFwvBQ"
                      font-size="12px"
                      height="19px"
                      width="15px"
                    >
                      +
                    </button>
                  </div>
                </div>
                <p
                  class="sc-dkrFOg dNFUVo sc-jSUZER QifgC"
                  font-size="14px"
                  font-weight="700"
                >
                  R$19
                </p>
              </div>
            </div>
            <button
              class="sc-gswNZR cnmGuK"
              height="18px"
              transform="translate(10px, -45px)"
              width="18px"
            >
              <p
                class="sc-dkrFOg harcBR"
                color="#FFFFFF"
                font-size="8px"
              >
                X
              </p>
            </button>
          </div>
        </div>
      </div>
      <div
        class="sc-iBYQkv tvFcD"
        width="100%"
      >
        <div
          class="sc-iBYQkv biqZAK"
          width="80%"
        >
          <p
            class="sc-dkrFOg kYzYoE"
            color="#FFFFFF"
            font-size="28px"
            font-weight="700"
          >
            Total:
          </p>
          <p
            class="sc-dkrFOg kYzYoE"
            color="#FFFFFF"
            font-size="28px"
            font-weight="700"
          >
            R$29.98
          </p>
        </div>
        <button
          class="sc-gswNZR cUQGLN"
          color="#FFFFFF"
          font-size="28px"
          height="97px"
          width="100%"
        >
          <p
            class="sc-dkrFOg cxIDJR"
            font-weight="700"
          >
            Finalizar Compra
          </p>
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
